pkg/query-service/app/clickhouseReader/reader.go

```diff
@@ -2753,3 +2762,289 @@ func (r *ClickHouseReader) GetSamplesInfoInLastHeartBeatInterval(ctx context.Con
+		"CAST((attributes_float64_key, attributes_float64_value), 'Map(String, Float64)') as  attributes_float64,"+
+		"CAST((resources_string_key, resources_string_value), 'Map(String, String)') as resources_string "+
+		"from %s.%s", r.logsDB, r.logsTable)
+
+	if filterSql != nil && *filterSql != "" {
+		query += fmt.Sprintf(" where %s", *filterSql)
+	}
+
+	query = fmt.Sprintf("%s order by %s %s limit %d", query, params.OrderBy, params.Order, params.Limit)
+	zap.S().Debug(query)
+	err = r.db.Select(ctx, response, query)
+	if err != nil {
+		return nil, &model.ApiError{Err: err, Typ: model.ErrorInternal}
+	}
+
+	return response, nil
+}
+
+func (r *ClickHouseReader) TailLogs(ctx context.Context, client *model.LogsTailClient) {
+
+	fields, apiErr := r.GetLogFields(ctx)
+	if apiErr != nil {
+		client.Error <- apiErr.Err
+		return
+	}
+
+	filterSql, err := logs.GenerateSQLWhere(fields, &model.LogsFilterParams{
+		Query: client.Filter.Query,
+	})
+
+	if err != nil {
+		client.Error <- err
+		return
+	}
+
+	query := fmt.Sprintf("SELECT "+
+		"timestamp, id, trace_id, span_id, trace_flags, severity_text, severity_number, body,"+
+		"CAST((attributes_string_key, attributes_string_value), 'Map(String, String)') as  attributes_string,"+
+		"CAST((attributes_int64_key, attributes_int64_value), 'Map(String, Int64)') as  attributes_int64,"+
+		"CAST((attributes_float64_key, attributes_float64_value), 'Map(String, Float64)') as  attributes_float64,"+
+		"CAST((resources_string_key, resources_string_value), 'Map(String, String)') as resources_string "+
+		"from %s.%s", r.logsDB, r.logsTable)
+
+	currentTime := uint64(time.Now().UnixNano() / int64(time.Millisecond))
+	tsStart := &currentTime
+	if client.Filter.TimestampStart != nil {
+		tsStart = client.Filter.TimestampStart
+	}
+
+	var idStart *string
+	if client.Filter.IdStart != nil {
+		idStart = client.Filter.IdStart
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			done := true
+			client.Done <- &done
+			zap.S().Debug("closing go routine : " + client.Name)
+			return
+		default:
+			tmpQuery := fmt.Sprintf("%s where timestamp >='%d'", query, *tsStart)
+			if filterSql != nil && *filterSql != "" {
+				tmpQuery += fmt.Sprintf(" and %s", *filterSql)
+			}
+			if idStart != nil {
+				tmpQuery += fmt.Sprintf(" and id > '%s'", *idStart)
+			}
+			tmpQuery = fmt.Sprintf("%s order by timestamp asc, id asc limit 1000", tmpQuery)
+			zap.S().Debug(tmpQuery)
+			response := &[]model.GetLogsResponse{}
+			err := r.db.Select(ctx, response, tmpQuery)
+			if err != nil {
+				zap.S().Debug(err)
+				client.Error <- err
+				return
+			}
+			len := len(*response)
+			for i := 0; i < len; i++ {
+				select {
+				case <-ctx.Done():
+					done := true
+					client.Done <- &done
+					zap.S().Debug("closing go routine while sending logs : " + client.Name)
+					return
+				default:
+					client.Logs <- &(*response)[i]
+					if i == len-1 {
+						tsStart = &(*response)[i].Timestamp
+						idStart = &(*response)[i].ID
```

Comment:
Same comment as above about de-referencing, Is there a reason why you are doing this get address then deref instead of just using the value itself?

Reply:
Sure have corrected this and made it consistent